### ISSUE TITLE
Set retry-after http header on retryable responses

### ISSUE THREAD

**Issue by bboreham** [grafana/mimir#5233]
#### Is your feature request related to a problem? Please describe.

Prometheus remote-write defaults to retry very quickly.  Especially when there are a lot of agent shards retrying in parallel, this can overwhelm the server.

#### Describe the solution you'd like

Mimir should set the `retry-after` header, which is honoured by Prometheus.  E.g. it can set it to 5 seconds from now, so the agent waits 5 seconds.

Haven't thought about this deeply but Mimir can see what the recent rate of requests from this tenant has been, and perhaps compute what length of time the client should wait to bring the rate below the configured rate limit.  Maybe also look at the number of inflight requests.

#### Describe alternatives you've considered

Doing nothing.

**Comment by mattmendick** [comment]
Great idea 👍 can this header be set on other http status codes? Such as 5xx, since they will be retried by remote-write, and the system might benefit from the senders slowing down a bit.

**Comment by noise** [comment]
Agent to be updated to honor for non-429s: https://github.com/grafana/agent/issues/4652

**Comment by mattmendick** [comment]
I _think_ that code might just be for the agent management config. Remote write logic I think is provided by https://github.com/prometheus/prometheus/blob/4a56a6bf592ad032714ddedaee1b5e4fc77aa93e/storage/remote/client.go#L238

**Comment by noise** [comment]
Upstream Prometheus issue for the same: https://github.com/prometheus/prometheus/issues/12674

**Comment by amurray2306** [comment]
Hey there was this resolved or is it still open?

**Comment by jhalterman** [comment]
We should also factor in `X-Retry-Count` when computing a retry-after duration, as this is being added upstream to the agent and prometheus:

https://github.com/grafana/agent/issues/4650
https://github.com/prometheus/prometheus/pull/12729

**Comment by ying-jeanne** [comment]
The PR of Bryan on prometheus: https://github.com/prometheus/prometheus/pull/12677
PR on agent: https://github.com/grafana/agent/pull/5152

**Comment by ying-jeanne** [comment]
A design doc for setting retry-after header https://docs.google.com/document/d/1O1bxZBTH8_kXLefNP936rRqj14vJ5jaLgg4FeqAO0is in discussion

**Comment by mattmendick** [comment]
The agent team has been ushering the X-Retry-Count and Retry-After issues through (both been merged into Prometheus), and will be in an agent soon (~3 weeks until next release, Around October 11), so all the requirements are coming together nicely for this to work out 👍

**Comment by pstibrany** [comment]
`Retry-Attempt` header was added in https://github.com/prometheus/prometheus/pull/12729

**Comment by mattmendick** [comment]
Just to push an update here. In response to [Dee asking about it on slack](https://raintank-corp.slack.com/archives/C029912SXT8/p1698921269574999), @ying-jeanne said: 
> I am still testing this one. Not deployed to WF yet. Just sent a message to Andy. I plan to finish the test and merge code change next week.

So targeting early December for delivery.

**Comment by dimitarvdimitrov** [comment]
On Oct 17 @mattmendick wrote

@ying-jeanne How's this one looking, are you still working on it? This would be really helpful as the latest agent has the support for retry-after on non-2xx requests, so we would be keen to enable this.

Even if we can't, having this support for other users who do upgrade to the agent or prometheus will be very helpful.

**Comment by dimitarvdimitrov** [comment]
On Nov 6 @mattmendick wrote

Just to push an update here. @ying-jeanne said: 
> I am still testing this one. Not deployed yet. Just sent a message to Andy. I plan to finish the test and merge code change next week.

So targeting early December for delivery.

**Comment by ying-jeanne** [comment]
Hi, just to give a quick update. I am testing it today on mimir-dev-10. Plan to finish the test and merge the code this week. Target early December for delivery, the status didn't change since last time.

**Comment by ying-jeanne** [comment]
Load test is done on dev and Pull request is ready for review, pending for review.

**Comment by ying-jeanne** [comment]
implemented and activated for all cells
